v1/remote/image.go

```diff
@@ -16,19 +16,125 @@ package remote
 
 import (
 	"fmt"
+	"io"
+	"io/ioutil"
 	"net/http"
+	"net/url"
+	"sync"
 
 	"github.com/google/go-containerregistry/authn"
 	"github.com/google/go-containerregistry/name"
 	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/partial"
+	"github.com/google/go-containerregistry/v1/remote/transport"
+	"github.com/google/go-containerregistry/v1/types"
 )
 
-// TODO(jonjohnson): This.
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+// image accesses an image from a remote registry
+type image struct {
+	ref    name.Reference
+	client *http.Client
+}
+
+type compressedImage struct {
+	*image
+	manifestLock sync.Mutex // Protects manifest
+	manifest     []byte
+	configLock   sync.Mutex // Protects config
+	config       []byte
+}
+
+var _ partial.CompressedImageCore = (*compressedImage)(nil)
 
 // Image accesses a given image reference over the provided transport, with the provided authentication.
 func Image(ref name.Reference, auth authn.Authenticator, t http.RoundTripper) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: remote.Image(%v)", ref)
+	tr, err := transport.New(ref, auth, t, transport.PullScope)
+	if err != nil {
+		return nil, err
+	}
+	img := image{
+		ref: ref,
+		client: &http.Client{
+			Transport: tr,
+		},
+	}
+	c := compressedImage{
+		image: &img,
+	}
+	return partial.CompressedToImage(&c)
```

Comment:
I'm really happy about how tight this code ended up being.

Reply:
Same :+1: